app/admin/resources/validators/utils.ts

```diff
@@ -0,0 +1,35 @@
+import vine, { errors } from "@vinejs/vine";
+import { ActionRequest, ValidationError } from "adminjs";
+
+interface VineError {
+  message: string;
+  rule: string;
+  field: string;
+}
+
+export const validateResource = async (
+  validator: ReturnType<typeof vine.compile>,
+  request: ActionRequest,
+) => {
+  const { method, payload } = request;
+  if (method === "post" && payload !== undefined) {
+    try {
+      await validator.validate(payload);
+    } catch (err) {
+      if (err instanceof errors.E_VALIDATION_ERROR) {
+        const errorMessages = err.messages as VineError[];
+        const errorsMap = errorMessages.reduce(
+          (acc, { message, field, rule }) => ({
+            ...acc,
+            [field]: { message, type: rule },
+          }),
+          {},
+        );
+        throw new ValidationError(errorsMap, {
+          message: "Oops you have validation errors",
+        });
+      }
+    }
+    return request;
+  }
+};
```

Comment:
don't use arrow functions for long exported functions
also explicitly type the return type, otherwise you might miss a return statement on some branch